doc/design/drafts/dynamic_size_types.md

```diff
@@ -0,0 +1,433 @@
+overhead. One example could be that someone analyses an image with a function
+like
+`void analyseImage(cxx::vector<int, SharedMemoryAllocator> & a)`. This function
+can now be only used with a vector which is using the `SharedMemoryAllocator`.
+One could define the allocator type as a template but this would increase the
+compile time and can make it nearly impossible to write proprietary code since
+templates require the implementation in the header. Furthermore, the function
+developer has to be aware of the implementation detail that there is something
+like a `SharedMemoryAllocator`.
+
+## Memory Layout
+
+In the following section we would like to send the structure 
+
+```cpp
+struct MyData {
+  int         value;
+  vector<int> image;
+  string      text;
+};
+```
+
+and observe how this is stored currently in the shared memory of iceoryx and
+compare it to the suggested solution.
+
+An iceoryx containers memory structure looks mostly like this:
+```
+vector/string {
+  mgmt; // contains internal variables to handle the logic and later the
+        // suggested allocator
+
+  data; // contains the user data which is stored in the container
+}
+```
+
+### Static Layout
+
+When we require a sample with our publisher and write `MyData` we gain have a
+chunk where all the contents are stored inside one chunk. Hereby is the
+Memory Manager the class which handles the distribution of shared memory chunks.
+To support different sizes the Memory Manager uses bucket allocators with
+preconfigured bucket sizes called Memory Pool.
+
+```
++--------------------- Memory Manager -------------------+
+|  MemPool with chunk size = 1000                        |
+|  +----------------+----------------+----------------+  |
+|  |  data          |                |                |  |
+|  |  image::mgmt   |                |                |  |
+|  |  image::data   |                |                |  |
+|  |  text::mgmt    |                |                |  |
+|  |  text::data    |                |                |  |
+|  +----------------+----------------+----------------+  |
+|                                                        |
+|  MemPool with chunk size = 100                         |
+|  +---+---+---+---+---+---+---+---+---+---+---+---+---+ |
+|  |   |   |   |   |   |   |   |   |   |   |   |   |   | |
+|  |   |   |   |   |   |   |   |   |   |   |   |   |   | |
+|  |   |   |   |   |   |   |   |   |   |   |   |   |   | |
+|  +---+---+---+---+---+---+---+---+---+---+---+---+---+ |
++--------------------------------------------------------+
+```
+
+### Layout With Dynamic Sized Types
+
+With dynamic sized types we move the data part outside of the container and
+add an allocator to the container which allows the container to acquire
+memory for their data part. When we assume that `MyData::image` and
+`MyData::text` require 100 bytes for their data part the memory layout could
+look like this.
```

Comment:
What happens when there is not enough memory for the data Part?

Reply:
I adjusted the allocator interface so that it provides the memory packed inside an expected with a custom allocator error type. This is error is forwarded via `reserve` when it fails.